impute_constant/impute_constant.tex

```diff
@@ -0,0 +1,158 @@
+        return list(map(impute_constant, data)) |\label{line:map}|
+        
+    return Transformation(input_domain, output_domain, function(data?), input_metric, output_metric, stability_relation=Relation)
+
+\end{lstlisting}
+
+\grace{ Will need to change pseudocode so that it returns the result of a make row by row transformation (which the code does) instead of a Transformation directly. Make sure to ask.}
+
+\subsection{Proof}
+\begin{theorem}
+
+
+For every setting of the input parameters \texttt{constant} to \texttt{make\_impute\_constant} such that the given preconditions hold, the transformation returned by \texttt{make\_impute\_constant} has the following properties:
+\begin{enumerate}
+    \item \textup{(Appropriate output domain).} If vector $v$ is in the \texttt{input\_domain}, then \texttt{function(v)} is in the \texttt{output\_domain}.
+    \item \textup{(Domain-Metric Compatibility).} The domain \texttt{input\_domain} matches one of the possible domains listed in the definition of \texttt{input\_metric}, and likewise \texttt{output\_domain} matches one of the possible domains listed in the definition of \texttt{output\_metric}.
+    \item \textup{(Stability Guarantee).} For every pair of elements $v, w$ in \texttt{input\_domain} and for every pair $(\din, \dout)$, where $\din$ is of the associated type for \texttt{input\_metric} and $\dout$ is the associated type for \texttt{output\_metric}, if $v,w$ are $d_{in}$-close under \texttt{input\_metric} and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$ are $d_{out}$-close under \texttt{output\_metric}.
+\end{enumerate}
+\end{theorem}
+\begin{proof}
+\begin{enumerate}
+\item \textbf{(Appropriate output domain).} 
+In the case of \texttt{make\_impute\_constant}, this corresponds to showing that for every vector $v$ of elements of type \texttt{DA}, \texttt{function}$(v)$ is a vector of elements of type \texttt{DA::NonNull}.
+
+The $\function(v)$ has type \texttt{Vec(DA)} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(DA)} and returns an element of type \texttt{Vec(DA::NonNull)}. If the Rust code compiles correctly, then the type correctness follows from the definition of the type signature enforced by Rust. Otherwise, the code raises an exception for incorrect input type. 
+
+Lastly, we ensure that return type must be \texttt{NonNull} \grace{Attribute? Property? Carrier} \texttt{DA::NonNull} because we check to make sure the \texttt{constant} being imputed in the function must be null. \grace{Not sure if this needs to be explicit in the pseudo code with the error check, or if I can just cite some trait property.} This check already exists because in the input of the \texttt{make\_impute\_constant}, \texttt{constant} is required to have type of \texttt{DA::NonNull}.
```

Comment:
Want feedback on appropriate output domain - not sure how to make it more rigorous, and if I need to consider more than just the type signature.

Reply:
[Based on this message](https://github.com/opendp/whitepapers/pull/8#issuecomment-889511572) I think there should first be a definition of AllDomain<T> along the lines of "the set of all non-null values that T may take on." You want to show that the output domain on this constructor is `VectorDomain<AllDomain<T>>`, or "any `Vec<T>` without nulls". You can first lean on the type system to get you most of the way there- the type system constrains the output set to "any `Vec<T>`". You can then show that the set of potential outputs is equivalent to the set described by `VectorDomain<AllDomain<T>>` by showing that vectors with nulls are not members of the output set.

Reply:
Unmarking this as resolved, because you resolved it before I finished writing my response!

Reply:
The Ord changes make this comment is partially outdated. AllDomain will not be adjusted.